### ISSUE TITLE
Playlist songs are re-sorted after play commences

### ISSUE THREAD

**Issue by zeezhao** [quodlibet/quodlibet#3957]
Steps to reproduce
------------------
Switch to a playlist and start playing first track.

Expected Output
---------------
Quodlibet should play track, and order of playlist should be maintained.

Actual Output
-------------
Quodlibet starts playing, but automatically sorts the playlist based on Title. Hence order of playlist not maintaind.

Test System
-----------

### Which version of Quod Libet?
4.5.0

### Which operating system 
Windows 10

### If it's audio-related, what back-end?

Additional Information
----------------------
This was perfect in v4.4.0.
The only reason I upgraded is because of https://github.com/quodlibet/quodlibet/issues/3579

**Comment by zeezhao** [comment]
I reverted back to previous version v4.4.0 but resinstalled v4.5.0 again just to recheck. It seems v4.5.0 is sorting playlists based on "Track" and/or "Title" if track is empty. This should not be the case as playlists can be from multiple albums and should always follow user-defined order. Please revert to no default sorting on playlists. Thanks.
For now, I have to go back to v4.4.0.

**Comment by declension** [comment]
Thanks for filling. There are definitely some bugs with the playlist browser, but I can't replicate this described behaviour (on Linux). 

There is no default sorting on playlists; in fact the songlist is shared across browsers. It reacts to clicking column headings to indicate sort preference in the songs - but typically gives you a visual indication of this sorting with an arrow (like in most Gnome apps).

Have you got _Sort Songs when tags are modified_ under _Preferences_ -> _Behaviour_ set? That seems closest and is relatively new (see #2509), though again seems fine for me.

**Comment by zeezhao** [comment]
Thanks for your reply.
I had "Sort Songs when tags are modified" set. But I unset it and it gave exactly the same situation. 

The columns are sorted by "Track" column.

To replicate behaviour:
- Go to a playlist.
- Double-click the first or any track in the playlist to start playing.
- The playlist then gets automatically sorted by "Track" and plays the selected track.

Thanks for looking into this.

**Comment by declension** [comment]
> To replicate behaviour:
> 
>     * Go to a playlist.
> 
>     * Double-click the first or any track in the playlist to start playing.
> 
>     * The playlist then gets automatically sorted by "Track" and plays the selected track.

Yep - this doesn't happen for me though. Perhaps it's a Windows thing, best I can think of right now

**Comment by zeezhao** [comment]
The fact that it worked perfectly in 4.4.0 and all previous versions, makes me doubt it is a Windows issue though. 

Is there any change in 4.5.0 that you believe could give rise to platform-specific issues? I had a look at the changelist and I am not sure of any.

**Comment by zeezhao** [comment]
Or an unintended change from #2509 affecting playlists?

**Comment by FerdiScholten** [comment]
I have exactly the same issue on Ubuntu Linux 21.10
Makes version 4.5.0 unusable for me. My playlists for my lessons need to remain in exactly the order as created. I will revert back to 4.4.0 that worked flawless.

**Comment by FerdiScholten** [comment]
When selecting another playlist and then going back to the autosorted playlist, the original order is restored. The when the next song starts to play the list is again being autosorted.

**Comment by zeezhao** [comment]
@FerdiScholten 
@FerdiScholten - I have this issue too. I should have also mentioned this. Thanks.
I have had to revert back to 4.4.0 too.

> When selecting another playlist and then going back to the autosorted playlist, the original order is restored. The when the next song starts to play the list is again being autosorted.

**Comment by FerdiScholten** [comment]
https://user-images.githubusercontent.com/35849405/161860168-ff3d0a12-73fd-4470-9942-a52f8a9cebab.mp4

A simple video might explain more, here is a quick overview of what is happening with my playlists.

**Comment by Meriipu** [comment]
Have you enabled any plugins?

Does it happen if you disabled the option introduced in #2509 as well as potentially disabling "jump to playing song automatically" ?

So both `Jump to playing song automatically` and `Sort songs when tags are modified` ?

**Comment by slosd** [comment]
I can reproduce this, though not after https://github.com/quodlibet/quodlibet/pull/2509. Git bisect points me to this commit https://github.com/quodlibet/quodlibet/commit/bc03a5a6393ccef53b280852051f3ffee5127594

There are two issues as far as I can see:

1. Try to drag a song to a new position -> The list is not updated, only after switching to another playlist and then back.
2. Double click a song to start playing -> The list turns on sorting (presumably with the last known sorting headers).

**Comment by FerdiScholten** [comment]
> Have you enabled any plugins?
> 
> Does it happen if you disabled the option introduced in #2509 as well as potentially disabling "jump to playing song automatically" ?
> 
> So both `Jump to playing song automatically` and `Sort songs when tags are modified` ?

Yes I have both options disabled. It does not make any difference.

**Comment by zeezhao** [comment]
> Have you enabled any plugins?
> 
> Does it happen if you disabled the option introduced in #2509 as well as potentially disabling "jump to playing song automatically" ?
> 
> So both `Jump to playing song automatically` and `Sort songs when tags are modified` ?

Disabling both options gives the same bug (on Windows 10).

The plugins I always have set are: Advanced Preferences, Alternative Seek Bar, Application Information, Change Language, Conditional Query, Convert Encodings, Duplicates Browser, Equaliser, Export Playlist to Folder, Export as M3U/PLS Playlist File, Export to HTML, Include Saved Search, Rescan Songs, Theme Switcher, Waveform Seek Bar.

**Comment by CeraRoseMadrigal** [comment]
Just upgraded to Quod Libet 4.5.0, on Pop! OS 20.04 LTS, and running into this same issue. When I first open Quod Libet, my playlists are shown in the correct order that I put the tracks in, but when I go to play any of them they seem to resort based on album title. No arrow indicates any particular sort column as being 'active', and clicking and dragging to rearrange the playlist no longer works. When I close and re-open Quod Libet, the playlist is back in the order it originally was, but resorts again if I try to play it.

Plugins installed: Advanced Preferences, Application Information, Change Language, Filter on Any Tag, MPRIS D-Bus Support, Theme Switcher, View Lyrics, Waveform Seek Bar.

**Comment by FerdiScholten** [comment]
In order to make sure that other plugins are not causing this, I made a backup of my old config.
Then I removed the Quodlibet configuration and started Quodlibet. This was as expected a completely clean startup. No plugins enabled. No music, nothing...
I let Quodlibet recreate my library and imported my playlists after that.
This resulted unfortunately in exactly the same behavior. Auto sorting the playlist when starting playback and/or on progressing to the next song. This is without any plugins loaded, so it rules out a plugin problem.

**Comment by declension** [comment]
Thanks for all the investigations and esp @slosd for the bisecting. Obviously was hoping it wouldn't be that monster commit, but hey...

Very short of time in the next few weeks but I'll push a branch with ideas (since I can't easily replicate behaviour) and maybe people who are comfortable can pull this and see what works...

Longer term we need _way_ more (any?) tests around playlist / songslist interaction but they can be pretty tricky, given how global and deep into GTK+ that code is

**Comment by declension** [comment]
@slosd @FerdiScholten @zeezhao can anyone pull that branch and test? :pray:

**Comment by zeezhao** [comment]
> @slosd @FerdiScholten @zeezhao can anyone pull that branch and test? 🙏

Hi. Unfortunately, I am not setup to test this via github.... But if you send me a link to a release with the change I can setup on Windows 10 and test. Thanks for your help on the bug.

**Comment by FerdiScholten** [comment]
Just tested and the behavior did not change. The list still gets sorted at double click or at the end of the playing song.

**Comment by FerdiScholten** [comment]
And debug output when ending a song from the playlist and it gets sorted after ending.
D: 23.685: player.gstbe.player.__about_to_finish: About to finish (async)
D: 23.685: player.gstbe.player.__about_to_finish_sync: About to finish (sync)
D: 23.685: player.gstbe.player.__about_to_finish_sync: Select next song in mainloop..
D: 23.685: PlaylistModel.next_ended: Using <One song ∘ In Order>.next_implicit() to get next song
D: 23.685: OneSong.next: Ending songlist.
D: 23.685: player.gstbe.player.__about_to_finish_sync: ..done.
D: 23.685: player.gstbe.player.__about_to_finish: About to finish (async): done
D: 24.967: player.gstbe.player.__message: Stream EOS
D: 24.967: GStreamerPlayer._end: End song
D: 24.969: player.gstbe.player.__destroy_pipeline: Destroying Gstreamer pipeline
D: 24.972: RandomAlbum.plugin_on_song_started: Playlists doesn't support album filtering
D: 24.976: SongFileLibrary._changed: Emitting changed for 1 item(s) (e.g. '/home/ferdi/Music/Biodanza/HLB/HLBC27 - 64 - Rada, Rubén - Quien va a cantar.mp3'...) from <SongFileLibrary @ 0x7fb869763380>
D: 24.977: qltk.completion.__update_song: Updated tag model for 1 songs
D: 24.977: XSPFBackedPlaylist.write: Writing '/home/ferdi/.config/quodlibet/playlists/2022-04-19 Leven is bewegen.xspf'
D: 24.978: PlaylistLibrary.changed('PlaylistLibrary for main'): Changing 1 items directly.
D: 24.978: PlaylistLibrary._changed: Emitting changed for 1 item(s) (e.g. '2022-04-19 Leven is bewegen'...) from <PlaylistLibrary @ 0x7fb85f8d8540>
D: 24.978: PlaylistsBrowser.changed: Refreshing view in <PlaylistsBrowser @ 0x7fb85f8c9100> for <XSPFBackedPlaylist at '/home/ferdi/.config/quodlibet/playlists/2022-04-19 Leven is bewegen.xspf'>
D: 24.978: PlaylistsBrowser.changed: Updating songslist for selected <XSPFBackedPlaylist at '/home/ferdi/.config/quodlibet/playlists/2022-04-19 Leven is bewegen.xspf'>
D: 24.979: PlaylistModel.set: Filling view model with 14 songs.

**Comment by FerdiScholten** [comment]
Made some video's again. Now showing the debug output, hope that helps. Looks like the double click action and the end of song action leading to the sorting are different.
sorting in this video happen 2 times:
first after 4,266 seconds after loading the playlist on startup
second after 25,600 seconds after double click to start the song
Both show **qltk.tracker.__changed: Current song changed, updating current file** in the debug output during this sort

https://user-images.githubusercontent.com/35849405/164317372-fba52d10-9501-4ae6-a416-f2cbbbdbb7f8.mp4

**Comment by FerdiScholten** [comment]
Now for the end of song sorting.
In this video the sorting happens after 20,583 seconds.
Debug output shows **Playlistmodel.set: Filling view model with 16 songs** during the sort.

Hope this helps.

https://user-images.githubusercontent.com/35849405/164319243-a4c52d07-148a-408d-b610-e0d787a4b361.mp4

**Comment by FerdiScholten** [comment]
**Please ignore what I posted above, I made a mistake and it is still using the main branch.**

Now correctly switched to the playlist-reordering-fixes branch an can confirm that the reordering issue is fixed.
My sincere apologies for the inconvenience...

**Comment by asger-finding** [comment]
This my only nitpick about what I consider to be _the best_ music player for Linux (and maybe the best as a whole)

I'll continue to use the program, it's just super annoying that I have to disable the Track header whenever I want to listen to my playlists.

**Comment by zeezhao** [comment]
Unfortunately, I have stuck with version 4.4.0 until the fix is migrated. 

Based  on previous comments, if you can build latest version from repository, the fix works.

**Comment by pascalspits** [comment]
... any idea when this "fix" will be migrated into the main version ?

**Comment by FerdiScholten** [comment]
This has been fixed 9 months ago, but still the 4.6 release remains elusive.
